tests/Database/DatabaseConnectionTest.php

```diff
@@ -148,18 +151,42 @@ public function testTransactionLevelNotIncrementedOnTransactionException()
         }
     }
 
-    public function testBeginTransactionMethodRetriesOnFailure()
+    public function testBeginTransactionMethodUsesCreateTransactionMethod()
     {
         $pdo = $this->createMock(DatabaseConnectionTestMockPDO::class);
-        $pdo->expects($this->at(0))
-            ->method('beginTransaction')
-            ->will($this->throwException(new ErrorException('server has gone away')));
-        $connection = $this->getMockConnection(['reconnect'], $pdo);
-        $connection->expects($this->once())->method('reconnect');
+        $connection = $this->getMockConnection(['createTransaction', 'fireConnectionEvent'], $pdo);
+        $connection->expects($this->once())->method('createTransaction');
+        $connection->expects($this->once())->method('fireConnectionEvent')->with('beganTransaction');
         $connection->beginTransaction();
         $this->assertEquals(1, $connection->transactionLevel());
     }
 
+    public function testCreateTransactionMethodRetriesOnFailure()
+    {
+        $pdo = $this->createMock(DatabaseConnectionTestMockPDO::class);
+        $exception = new ErrorException('server has gone away');
+        $pdo->expects($this->once())
+            ->method('beginTransaction')
+            ->will($this->throwException($exception));
+        $requiredMethods = ['reconnectIfMissingConnection', 'handleBeginTransactionException'];
+        $connection = $this->getMockConnection($requiredMethods, $pdo);
+        $connection->expects($this->once())->method('reconnectIfMissingConnection');
+        $connection->expects($this->once())->method('handleBeginTransactionException')->with($exception);
+        $this->privateCall($connection, 'createTransaction');
+    }
+
+    public function testHandleBeginTransactionExceptionRetries()
+    {
+        $exception = new ErrorException('server has gone away');
+        $pdo = $this->createMock(DatabaseConnectionTestMockPDO::class);
+        $requiredMethods = ['causedByLostConnection', 'reconnect'];
+        $connection = $this->getMockConnection($requiredMethods, $pdo);
+        $connection->expects($this->once())->method('causedByLostConnection')->with($exception)->willReturn(true);
+        $connection->expects($this->once())->method('reconnect');
+        $pdo->expects($this->once())->method('beginTransaction');
+        $this->privateCall($connection, 'handleBeginTransactionException', [$exception]);
+    }
+
```

Comment:
You seem to have added new tests here. These are probably best done in a separate PR.

Reply:
They are because by removing `at` function there was a method which was not covered anymore. These tests are covering that method.Do you think they should be in separate PR?

Reply:
Yeah I think it'll be more clear to handle these in a separate PR with a good explanation why it's needed. Thanks

Reply:
Ok, i will try it.